File: 02 - Game of Life/nikhilmohan/python/gol.py
import time
import random

def emptyBoard(size):
    return [[0] * (size[1] + 2) for i in range(size[0] + 2)]

def randomBoard(size):
    board = emptyBoard(size)
    for y in range(len(board)-1):
        for x in range(len(board[0])-1):
            board[y+1][x+1] = random.randint(0, 1)
    return board

class World:
    def __init__(self, size, type='random'):
        self.size = size
        if type == 'random': b = randomBoard(size)
        else: b = emptyBoard(size)
        self.boards = [b, emptyBoard(size)]
        self.index = 0
        self.board = self.boards[self.index]

    def iterate(self):
        nb = self.boards[(self.index + 1) % 2]
        b = self.board

        r, c = self.size

        for y in range(1, r + 1):
            for x in range(1, c + 1):
                neighbors = b[y - 1][x - 1] + b[y - 1][x] + b[y - 1][x + 1] + b[y][x - 1] + b[y][x + 1] + b[y + 1][x - 1] + b[y + 1][x] + b[y + 1][x + 1]

                a = 0
                if b[y][x] == 0:
                    if neighbors == 3: a = 1
                else:
                    if neighbors == 2 or neighbors == 3: a = 1

                nb[y][x] = a

        self.index = (self.index + 1) % 2
        self.board = nb

if __name__ == "__main__":
    def f():
        w = World((1000, 1000))

        iterations = 10

        t = time.time()

        for i in range(iterations):
            w.iterate()

        t = time.time() - t
        t = 1000 * t / iterations

        print("%.2f" % t)

    f()

File: 02 - Game of Life/nikhilmohan/c/gol.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

#define N 1000

int boards[2][N+2][N+2], boardIndex = 0, (*board)[N+2], (*nb)[N+2];

void emptyBoard(int b[N+2][N+2]) {
  for(int y = 1; y <= N; y++)
    for(int x = 1; x <= N; x++)
      b[y][x] = 0;
}

void randomBoard(int b[N+2][N+2]) {
  for(int y = 1; y <= N; y++)
    for(int x = 1; x <= N; x++)
      b[y][x] = rand() % 2;
}

void iterate() {
  nb = boards[(boardIndex + 1) % 2];
  int x, y, a, neighbors = 0;

  for (y = 1; y <= N; y++) {
    for (x = 1; x <= N; x++) {
      neighbors = board[y-1][x-1] + board[y-1][x] + board[y-1][x+1] + board[y][x-1] + board[y][x+1] + board[y+1][x-1] + board[y+1][x] + board[y+1][x+1];

      a = 0;

      if (board[y][x] == 0) {
        if (neighbors == 3) a = 1;
      } else {
        if (neighbors == 2 || neighbors == 3) a = 1;
      }

      nb[y][x] = a;
    }
  }

  boardIndex = (boardIndex + 1) % 2;
  board = boards[boardIndex];
}

int main() {
  randomBoard(boards[0]);
  emptyBoard(boards[1]);
  board = boards[0];

  int iterations = 1000;

  struct timespec start, end;
  clock_gettime(CLOCK_MONOTONIC_RAW, &start);

  for(int i = 0; i < iterations; i++) iterate();

  clock_gettime(CLOCK_MONOTONIC_RAW, &end);

  int delta = (end.tv_sec - start.tv_sec) * 1000000 + (end.tv_nsec - start.tv_nsec) / 1000;
  double t = delta / (float) iterations / 1000;
  printf("%0.2f\n", t);

  return 0;
}
